python_modules/libraries/dagster-polars/dagster_polars/io_managers/type_routers.py

```diff
@@ -0,0 +1,173 @@
+            )
+        else:
+            return self.parent_type_router.is_eager
+
+    @property
+    def is_lazy(self) -> bool:
+        if self.is_root_type:
+            return issubclass(self.typing_type, pl.LazyFrame)
+        else:
+            return self.parent_type_router.is_lazy
+
+    def dump(self, obj: T, path: "UPath", dump_fn: F_D) -> None:
+        if self.is_root_type:
+            dump_fn(self.context, obj, path)
+        else:
+            self.parent_type_router.dump(obj, path, dump_fn)
+
+    def load(self, path: "UPath", load_fn: F_L) -> T:
+        if self.is_root_type:
+            return load_fn(path, self.context)
+        else:
+            return self.parent_type_router.load(path, load_fn)
+
+
+class TypeRouter(BaseTypeRouter, Generic[T]):
+    """Handles most common standard types."""
+
+    def match(self) -> bool:
+        return self.typing_type in [
+            pl.DataFrame,
+            pl.LazyFrame,
+            Any,
+            type(None),
+            None,
+        ]
+
+    @property
+    def is_root_type(self) -> bool:
+        return True
+
+
+class OptionalTypeRouter(BaseTypeRouter, Generic[T]):
+    """Handles Optional type annotations with a noop if the object is None or missing in storage."""
+
+    def match(self) -> bool:
+        return get_origin(self.typing_type) == Union and type(None) in get_args(self.typing_type)
+
+    @property
+    def is_root_type(self) -> bool:
+        return False
+
+    @property
+    def parent_type(self) -> Any:
+        return get_args(self.typing_type)[0]
+
+    def dump(self, obj: T, path: "UPath", dump_fn: F_D) -> None:
+        if obj is None:
+            self.context.log.warning(f"Skipping saving optional output at {path} as it is None")
+            return
+        else:
+            self.parent_type_router.dump(obj, path, dump_fn)
+
+    def load(self, path: "UPath", load_fn: F_L) -> T:
+        if not path.exists():
+            self.context.log.warning(f"Skipping loading optional input at {path} as it is missing")
+            return None
+        else:
+            return self.parent_type_router.load(path, load_fn)
+
+
+class DictTypeRouter(BaseTypeRouter, Generic[T]):
+    """Handles loading partitions as dictionaries of DataFrames."""
+
+    def match(self) -> bool:
+        return get_origin(self.typing_type) in (dict, Dict, Mapping)
+
+    @property
+    def is_root_type(self) -> bool:
+        return False
+
+    @property
+    def parent_type(self) -> Any:
+        return get_args(self.typing_type)[1]
+
+
+TYPE_ROUTERS = [
+    TypeRouter,
+    OptionalTypeRouter,
+    DictTypeRouter,
+]
+
+
+def resolve_type_router(
+    context: Union[InputContext, OutputContext], type_to_resolve: Any
+) -> TypeRouter:
+    """Finds the first matching TypeRouter for the given type."""
+    # try each router class in order of increasing complexity
+    for router_class in TYPE_ROUTERS:
+        router = router_class(context, type_to_resolve)
```

Comment:
I would prefer a staticmethod pattern here so we can just do `router_class.match(context, type_to_resolve)` rather than instantiating classes with invalid states

Reply:
Done